hedera-node/src/main/java/com/hedera/services/ledger/accounts/staking/RewardCalculator.java

```diff
@@ -20,66 +20,84 @@
  * ‍
  */
 
+import com.hedera.services.ledger.properties.AccountProperty;
 import com.hedera.services.state.merkle.MerkleAccount;
-import com.hedera.services.state.merkle.MerkleStakingInfo;
-import com.hedera.services.utils.EntityNum;
-import com.swirlds.merkle.map.MerkleMap;
 
 import javax.inject.Inject;
-import java.time.LocalDate;
-import java.time.ZoneId;
-import java.util.function.Supplier;
+import java.util.Map;
+
+import static com.hedera.services.ledger.accounts.staking.StakeChangeManager.finalBalanceGiven;
+import static com.hedera.services.ledger.accounts.staking.StakePeriodManager.isWithinRange;
+import static com.hedera.services.ledger.properties.AccountProperty.BALANCE;
+import static com.hedera.services.ledger.properties.AccountProperty.STAKE_PERIOD_START;
 
 public class RewardCalculator {
-	public static final EntityNum stakingFundAccount = EntityNum.fromLong(800L);
-	private final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts;
-	private final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfo;
+	private final StakePeriodManager stakePeriodManager;
+	private final StakeInfoManager stakeInfoManager;
 
-	public static final ZoneId zoneUTC = ZoneId.of("UTC");
+	private long rewardsPaid;
+	long accountReward;
+	long accountUpdatedStakePeriodStart;
 
 	@Inject
-	public RewardCalculator(final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts,
-			final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfo) {
-		this.accounts = accounts;
-		this.stakingInfo = stakingInfo;
+	public RewardCalculator(final StakePeriodManager stakePeriodManager,
+			final StakeInfoManager stakeInfoManager) {
+		this.stakePeriodManager = stakePeriodManager;
+		this.stakeInfoManager = stakeInfoManager;
 	}
 
-	public final long computeAndApplyRewards(final EntityNum accountNum) {
-		long todayNumber = LocalDate.now(zoneUTC).toEpochDay();
-		final var account = accounts.get().getForModify(accountNum);
+	public final void computePendingRewards(final MerkleAccount account) {
+		final long todayNumber = stakePeriodManager.currentStakePeriod();
 		var stakePeriodStart = account.getStakePeriodStart();
 
 		if (stakePeriodStart > -1 && stakePeriodStart < todayNumber - 365) {
-			account.setStakePeriodStart(todayNumber - 365);
+			stakePeriodStart = todayNumber - 365;
```

Comment:
Will avoid a bit of tech debt if we can write this like,
```
stakePeriodStart = Math.max(stakePeriodStart, stakePeriodManager.minimumValidStakePeriod());
```

Reply:
Made the changes